src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/student-exam-timeline.component.ts

```diff
@@ -0,0 +1,388 @@
+                const id = exercise.studentParticipations![0].id!;
+                submissionObservables.push(this.submissionService.findAllSubmissionsOfParticipation(id).pipe(map(({ body }) => body!)));
+            } else {
+                submissionObservables.push(
+                    this.submissionVersionService.findAllSubmissionVersionsOfSubmission(exercise.studentParticipations![0].submissions![0].id!).pipe(
+                        mergeMap((versions) => versions),
+                        toArray(),
+                    ),
+                );
+            }
+        });
+        return forkJoin(...submissionObservables);
+    }
+
+    /**
+     * Sorts the time stamps in ascending order
+     */
+    private sortTimeStamps() {
+        this.submissionTimeStamps = this.submissionTimeStamps.sort((date1, date2) => (date1.isAfter(date2) ? 1 : -1));
+    }
+
+    /**
+     * This method is called when the user clicks on the next or previous button in the navigation bar or on the slider.
+     * @param exerciseChange contains the exercise to which the user wants to navigate to and the submission that should be displayed
+     */
+    onPageChange(exerciseChange: {
+        overViewChange: boolean;
+        exercise?: Exercise;
+        forceSave: boolean;
+        submission?: ProgrammingSubmission | SubmissionVersion | FileUploadSubmission;
+    }): void {
+        const activeComponent = this.activePageComponent;
+        if (activeComponent) {
+            activeComponent.onDeactivate();
+        }
+
+        if (!exerciseChange.submission) {
+            // only change the submission if the exercise has changed
+            if (exerciseChange.exercise !== this.currentExercise) {
+                exerciseChange.submission = this.findSubmissionForExerciseClosestToCurrentTimeStampForExercise(exerciseChange.exercise!);
+            }
+        }
+
+        if (exerciseChange.submission) {
+            if (this.isSubmissionVersion(exerciseChange.submission)) {
+                const submissionVersion = exerciseChange.submission as SubmissionVersion;
+                this.selectedTimestamp = submissionVersion.createdDate.toDate().getTime();
+            } else if (this.isFileUploadSubmission(exerciseChange.submission)) {
+                const fileUploadSubmission = exerciseChange.submission as FileUploadSubmission;
+                this.selectedTimestamp = fileUploadSubmission.submissionDate!.toDate().getTime();
+            } else {
+                const programmingSubmission = exerciseChange.submission as ProgrammingSubmission;
+                this.selectedTimestamp = programmingSubmission.submissionDate!.toDate().getTime();
+            }
+            this.currentSubmission = exerciseChange.submission;
+            this.initializeExercise(exerciseChange.exercise!, exerciseChange.submission);
+        }
+    }
+
+    private findFirstSubmission(): FileUploadSubmission | SubmissionVersion | ProgrammingSubmission | undefined {
```

Comment:
An interface comment would be nice

Reply:
Why are only `FileUploadSubmission`s and `ProgrammingSubmission`s special? What about modeling and text exercises?

Reply:
modeling, text and quiz exercises are covered with the submission versions.
Submission versions do not exist for file upload exercises and programming exercises.